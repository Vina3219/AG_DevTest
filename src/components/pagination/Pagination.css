.pageNum--button {
  border: none;
  background-color: transparent;
  height: 40px;
  width: 40px;
  font-size: 16px;
}

.pageNum--button:disabled {
  background-color: white;
  color: black;
  font-style: bold;
  border-bottom: 4px solid #202020;
  text-align: center;
}

.pagination--button {
  border: none;
  background-color: transparent;
  font-size: 16px;
}
