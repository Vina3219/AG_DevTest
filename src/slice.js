import { createSlice } from '@reduxjs/toolkit';
import { loadProducts, getTotal } from './services'

const initialState = {
    products: [],
    pageNum: 1,
    productsCountPerPage: 8,
    totalProductsCount: 0
}

const slice = createSlice({
    name: 'app',
    initialState,
    reducers:{
        setPageNum: (state, action) => {
            state.pageNum = action.payload.pageNum;
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
        },
        setProductsPerPage: (state, action) => {
            state.productsCountPerPage = action.payload.productsCountPerPage;
            state.pageNum = Math.ceil(state.products[0].id/action.payload.productsCountPerPage);
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
        },
        loadPage: (state) => {
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
            state.totalProductsCount = getTotal();
        }
    }
})

export const { setPageNum, setProductsPerPage: setProductsCountPerPage, loadPage } = slice.actions;

export const selectProducts = (state) => state.app.products;
export const selectPageNum = (state) => state.app.pageNum;
export const selectTotalPage = (state) => Math.ceil(state.app.totalProductsCount/state.app.productsCountPerPage);
export const selectProductsCountPerPage = (state) => state.app.productsCountPerPage;
export const selectTotalProductsCount = (state) => state.app.totalProductsCount;

export default slice.reducer;
