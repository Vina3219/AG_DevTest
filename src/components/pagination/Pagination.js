import React, { useCallback, useMemo } from "react";
import { useSelector, useDispatch } from "react-redux";
import { selectPageNum, setPageNum, selectTotalPage } from "../../slice";
import "./Pagination.css";

const Pagination = () => {
  const pageNum = useSelector(selectPageNum);
  const dispatch = useDispatch();
  const handlePageNumChange = useCallback(
    (page) => (event) => {
      event.preventDefault();
      dispatch(
        setPageNum({
          pageNum: page,
        })
      );
    },
    [dispatch]
  );
  const totalPage = useSelector(selectTotalPage);

  const displayedPageNums = useMemo(() => {
    if (pageNum === 1) {
      if (totalPage <= 1) {
        return [1];
      }
      if (totalPage <= 2) {
        return [1, 2];
      }
      return [1, 2, 3];
    }
    if (pageNum === totalPage) {
      if (totalPage === 2) {
        return [1, 2];
      }
      return [pageNum - 2, pageNum - 1, pageNum];
    }
    return [pageNum - 1, pageNum, pageNum + 1];
  }, [pageNum, totalPage]);

  return (
    <>
      <button
        className="pagination--button"
        onClick={handlePageNumChange(pageNum - 1)}
        disabled={pageNum === 1}
      >
        &lt; Previous page
      </button>
      {displayedPageNums.map((page) => (
        <button
          className="pageNum--button"
          disabled={page === pageNum}
          key={page}
          onClick={handlePageNumChange(page)}
        >
          {page}
        </button>
      ))}
      <button
        className="pagination--button"
        onClick={handlePageNumChange(pageNum + 1)}
        disabled={pageNum === totalPage}
      >
        Next page &gt;
      </button>
    </>
  );
};

export default Pagination;
