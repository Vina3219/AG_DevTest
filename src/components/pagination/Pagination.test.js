import { useSelector, useDispatch } from "react-redux";
import { selectPageNum, setPageNum, selectTotalPage } from "../../slice";
import { fireEvent, render, screen } from "@testing-library/react";
import Pagination from "./Pagination";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock("../../slice", () => ({
  ...jest.requireActual("../../slice"),
  selectPageNum: jest.fn(),
  selectTotalPage: jest.fn(),
  setPageNum: jest.fn(),
}));

const dispatchSpy = jest.fn();

describe("Pagination", () => {
  beforeEach(() => {
    useSelector.mockReset().mockImplementation((selector) => selector());
    selectPageNum.mockReset().mockReturnValue(1);
    useDispatch.mockReset().mockReturnValue(dispatchSpy);
    selectTotalPage.mockReset().mockReturnValue(1);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Pagination />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays page number 1", () => {
    render(<Pagination />);
    const pageOneButton = screen.getByRole("button", { name: "1" });
    const previousButton = screen.getByRole("button", {
      name: /Previous page/,
    });
    const nextButton = screen.getByRole("button", {
      name: /Next page/,
    });
    expect(pageOneButton).toBeInTheDocument();
    expect(pageOneButton).toBeDisabled();
    expect(previousButton).toBeInTheDocument();
    expect(previousButton).toBeDisabled();
    expect(nextButton).toBeInTheDocument();
    expect(nextButton).toBeDisabled();
  });

  describe("if there are only two pages", () => {
    beforeEach(() => {
      selectPageNum.mockReset().mockReturnValue(1);
      selectTotalPage.mockReset().mockReturnValue(2);
    });

    it("matches snapshot", () => {
      const { asFragment } = render(<Pagination />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("displays page number 1 and 2", () => {
      render(<Pagination />);
      const pageOneButton = screen.getByRole("button", { name: "1" });
      const pageTwoButton = screen.getByRole("button", { name: "2" });
      const previousButton = screen.getByRole("button", {
        name: /Previous page/,
      });
      const nextButton = screen.getByRole("button", {
        name: /Next page/,
      });
      expect(pageOneButton).toBeInTheDocument();
      expect(pageOneButton).toBeDisabled();
      expect(pageTwoButton).toBeInTheDocument();
      expect(pageTwoButton).toBeEnabled();
      expect(previousButton).toBeInTheDocument();
      expect(previousButton).toBeDisabled();
      expect(nextButton).toBeInTheDocument();
      expect(nextButton).toBeEnabled();
    });

    describe("when the current page number is 2", () => {
      beforeEach(() => {
        selectPageNum.mockReset().mockReturnValue(2);
      });

      it("matches snapshot", () => {
        const { asFragment } = render(<Pagination />);
        expect(asFragment()).toMatchSnapshot();
      });

      it("displays page number 1 and 2", () => {
        render(<Pagination />);
        const pageOneButton = screen.getByRole("button", { name: "1" });
        const pageTwoButton = screen.getByRole("button", { name: "2" });
        const previousButton = screen.getByRole("button", {
          name: /Previous page/,
        });
        const nextButton = screen.getByRole("button", {
          name: /Next page/,
        });
        expect(pageOneButton).toBeInTheDocument();
        expect(pageOneButton).toBeEnabled();
        expect(pageTwoButton).toBeInTheDocument();
        expect(pageTwoButton).toBeDisabled();
        expect(previousButton).toBeInTheDocument();
        expect(previousButton).toBeEnabled();
        expect(nextButton).toBeInTheDocument();
        expect(nextButton).toBeDisabled();
      });
    });
  });

  describe("when there are 3 pages", () => {
    beforeEach(() => {
      selectPageNum.mockReset().mockReturnValue(1);
      selectTotalPage.mockReset().mockReturnValue(3);
    });

    it("matches snapshot", () => {
      const { asFragment } = render(<Pagination />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("displays page number 1 , 2 , 3", () => {
      render(<Pagination />);
      const pageOneButton = screen.getByRole("button", { name: "1" });
      const pageTwoButton = screen.getByRole("button", { name: "2" });
      const pageThreeButton = screen.getByRole("button", { name: "3" });
      const previousButton = screen.getByRole("button", {
        name: /Previous page/,
      });
      const nextButton = screen.getByRole("button", {
        name: /Next page/,
      });
      expect(pageOneButton).toBeInTheDocument();
      expect(pageOneButton).toBeDisabled();
      expect(pageTwoButton).toBeInTheDocument();
      expect(pageTwoButton).toBeEnabled();
      expect(pageThreeButton).toBeInTheDocument();
      expect(pageThreeButton).toBeEnabled();
      expect(previousButton).toBeInTheDocument();
      expect(previousButton).toBeDisabled();
      expect(nextButton).toBeInTheDocument();
      expect(nextButton).toBeEnabled();
    });

    describe("when current page is 2", () => {
      beforeEach(() => {
        selectPageNum.mockReset().mockReturnValue(2);
      });

      it("matches snapshot", () => {
        const { asFragment } = render(<Pagination />);
        expect(asFragment()).toMatchSnapshot();
      });

      it("displays page number 1 , 2 , 3", () => {
        render(<Pagination />);
        const pageOneButton = screen.getByRole("button", { name: "1" });
        const pageTwoButton = screen.getByRole("button", { name: "2" });
        const pageThreeButton = screen.getByRole("button", { name: "3" });
        const previousButton = screen.getByRole("button", {
          name: /Previous page/,
        });
        const nextButton = screen.getByRole("button", {
          name: /Next page/,
        });
        expect(pageOneButton).toBeInTheDocument();
        expect(pageOneButton).toBeEnabled();
        expect(pageTwoButton).toBeInTheDocument();
        expect(pageTwoButton).toBeDisabled();
        expect(pageThreeButton).toBeInTheDocument();
        expect(pageThreeButton).toBeEnabled();
        expect(previousButton).toBeInTheDocument();
        expect(previousButton).toBeEnabled();
        expect(nextButton).toBeInTheDocument();
        expect(nextButton).toBeEnabled();
      });

      describe("when clicks page 1", () => {
        it("dispatches setPageNum", () => {
          render(<Pagination />);
          const pageOneButton = screen.getByRole("button", { name: "1" });
          fireEvent.click(pageOneButton);
          expect(dispatchSpy).toBeCalledTimes(1);
          expect(setPageNum).toBeCalledTimes(1);
          expect(setPageNum).toBeCalledWith({
            pageNum: 1,
          });
        });
      });

      describe("when clicks previous page button", () => {
        it("dispatches setPageNum", () => {
          render(<Pagination />);
          const previousButton = screen.getByRole("button", {
            name: /Previous page/,
          });
          fireEvent.click(previousButton);
          expect(dispatchSpy).toBeCalledTimes(1);
          expect(setPageNum).toBeCalledTimes(1);
          expect(setPageNum).toBeCalledWith({
            pageNum: 1,
          });
        });
      });

      describe("when clicks next page button", () => {
        it("dispatches setPageNum", () => {
          render(<Pagination />);
          const nextButton = screen.getByRole("button", {
            name: /Next page/,
          });
          fireEvent.click(nextButton);
          expect(dispatchSpy).toBeCalledTimes(1);
          expect(setPageNum).toBeCalledTimes(1);
          expect(setPageNum).toBeCalledWith({
            pageNum: 3,
          });
        });
      });
    });

    describe("when current page is 3", () => {
      beforeEach(() => {
        selectPageNum.mockReset().mockReturnValue(3);
      });

      it("matches snapshot", () => {
        const { asFragment } = render(<Pagination />);
        expect(asFragment()).toMatchSnapshot();
      });

      it("displays page number 1 , 2 , 3", () => {
        render(<Pagination />);
        const pageOneButton = screen.getByRole("button", { name: "1" });
        const pageTwoButton = screen.getByRole("button", { name: "2" });
        const pageThreeButton = screen.getByRole("button", { name: "3" });
        const previousButton = screen.getByRole("button", {
          name: /Previous page/,
        });
        const nextButton = screen.getByRole("button", {
          name: /Next page/,
        });
        expect(pageOneButton).toBeInTheDocument();
        expect(pageOneButton).toBeEnabled();
        expect(pageTwoButton).toBeInTheDocument();
        expect(pageTwoButton).toBeEnabled();
        expect(pageThreeButton).toBeInTheDocument();
        expect(pageThreeButton).toBeDisabled();
        expect(previousButton).toBeInTheDocument();
        expect(previousButton).toBeEnabled();
        expect(nextButton).toBeInTheDocument();
        expect(nextButton).toBeDisabled();
      });
    });
  });
});
