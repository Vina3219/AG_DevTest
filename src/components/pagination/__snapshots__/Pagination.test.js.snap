// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination if there are only two pages matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="pagination--button"
    disabled=""
  >
    &lt; Previous page
  </button>
  <button
    class="pageNum--button"
    disabled=""
  >
    1
  </button>
  <button
    class="pageNum--button"
  >
    2
  </button>
  <button
    class="pagination--button"
  >
    Next page &gt;
  </button>
</DocumentFragment>
`;

exports[`Pagination if there are only two pages when the current page number is 2 matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="pagination--button"
  >
    &lt; Previous page
  </button>
  <button
    class="pageNum--button"
  >
    1
  </button>
  <button
    class="pageNum--button"
    disabled=""
  >
    2
  </button>
  <button
    class="pagination--button"
    disabled=""
  >
    Next page &gt;
  </button>
</DocumentFragment>
`;

exports[`Pagination matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="pagination--button"
    disabled=""
  >
    &lt; Previous page
  </button>
  <button
    class="pageNum--button"
    disabled=""
  >
    1
  </button>
  <button
    class="pagination--button"
    disabled=""
  >
    Next page &gt;
  </button>
</DocumentFragment>
`;

exports[`Pagination when there are 3 pages matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="pagination--button"
    disabled=""
  >
    &lt; Previous page
  </button>
  <button
    class="pageNum--button"
    disabled=""
  >
    1
  </button>
  <button
    class="pageNum--button"
  >
    2
  </button>
  <button
    class="pageNum--button"
  >
    3
  </button>
  <button
    class="pagination--button"
  >
    Next page &gt;
  </button>
</DocumentFragment>
`;

exports[`Pagination when there are 3 pages when current page is 2 matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="pagination--button"
  >
    &lt; Previous page
  </button>
  <button
    class="pageNum--button"
  >
    1
  </button>
  <button
    class="pageNum--button"
    disabled=""
  >
    2
  </button>
  <button
    class="pageNum--button"
  >
    3
  </button>
  <button
    class="pagination--button"
  >
    Next page &gt;
  </button>
</DocumentFragment>
`;

exports[`Pagination when there are 3 pages when current page is 3 matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="pagination--button"
  >
    &lt; Previous page
  </button>
  <button
    class="pageNum--button"
  >
    1
  </button>
  <button
    class="pageNum--button"
  >
    2
  </button>
  <button
    class="pageNum--button"
    disabled=""
  >
    3
  </button>
  <button
    class="pagination--button"
    disabled=""
  >
    Next page &gt;
  </button>
</DocumentFragment>
`;
